giraph-block-app/src/main/java/org/apache/giraph/block_app/framework/BlockUtils.java

```diff
@@ -147,6 +148,11 @@ public static void initAndCheckConfig(GiraphConfiguration conf) {
     checkBlockTypes(
         executionBlock, blockFactory.createExecutionStage(immConf), immConf);
 
+    PieceCount pieceCount = executionBlock.getPieceCount();
+    if (pieceCount.isKnown()) {
+      GiraphConstants.SUPERSTEP_COUNT.set(conf, pieceCount.getCount());
```

Comment:
There will be X+1 supersteps, but they are going to be supersteps 0..X. Actually I can make +1 here and then -1 in the logging part to keep it clear.